pkg/services/ngalert/image/service.go

```diff
@@ -0,0 +1,119 @@
+package image
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/grafana/pkg/components/imguploader"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/store"
+	"github.com/grafana/grafana/pkg/services/rendering"
+	"github.com/grafana/grafana/pkg/services/screenshot"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+//go:generate mockgen -destination=mock.go -package=image github.com/grafana/grafana/pkg/services/ngalert/image ImageService
+type ImageService interface {
+	// NewImage returns a new image for the alert instance.
+	NewImage(ctx context.Context, r *ngmodels.AlertRule, labels data.Labels) (*store.Image, error)
+}
+
+var (
+	// ErrNoDashboard is returned when the alert rule does not have a dashboard.
+	ErrNoDashboard = errors.New("no dashboard")
+
+	// ErrNoPanel is returned when the alert rule does not have a panel in a dashboard.
+	ErrNoPanel = errors.New("no panel")
+)
+
+const (
+	screenshotTimeout  = 10 * time.Second
+	screenshotCacheTTL = 15 * time.Second
+)
+
+// ScreenshotImageService takes screenshots of the panel for an alert rule and
+// saves the image in the store. The image contains a unique token that can be
+// passed as an annotation or label to the Alertmanager.
+type ScreenshotImageService struct {
+	screenshots screenshot.ScreenshotService
+	store       store.ImageStore
+}
+
+func NewScreenshotImageService(screenshots screenshot.ScreenshotService, store store.ImageStore) ImageService {
+	return &ScreenshotImageService{
+		screenshots: screenshots,
+		store:       store,
+	}
+}
+
+// NewScreenshotImageServiceFromCfg returns a new ScreenshotImageService
+// from the configuration.
+func NewScreenshotImageServiceFromCfg(cfg *setting.Cfg, metrics prometheus.Registerer,
+	db *store.DBstore, ds dashboards.DashboardService, rs rendering.Service) (ImageService, error) {
+	if !cfg.UnifiedAlerting.Screenshots.Enabled {
+		return &ScreenshotImageService{
+			screenshots: &screenshot.ScreenshotUnavailableService{},
+		}, nil
+	}
+
+	s := screenshot.NewBrowserScreenshotService(metrics, ds, rs)
+	if cfg.UnifiedAlerting.Screenshots.UploadExternalImageStorage {
+		u, err := imguploader.NewImageUploader()
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize uploading screenshot service: %w", err)
+		}
+		s = screenshot.NewUploadingScreenshotService(metrics, s, u)
+	}
+	s = screenshot.NewRateLimitScreenshotService(s, cfg.UnifiedAlerting.Screenshots.MaxConcurrentScreenshots)
+	s = screenshot.NewSingleFlightScreenshotService(s)
+	s = screenshot.NewCachableScreenshotService(metrics, screenshotCacheTTL, s)
+	s = screenshot.NewObservableScreenshotService(metrics, s)
```

Comment:
Nit(non-blocking): I think I would have done this differently, more go'ish, by doing something like

```golang
s := screenshot.NewService(withRateLimit(MaxConcurrentScreenshots), withCache(screenshotCacheTTL), withSingleFlight(),...)
````

Reply:
https://github.com/grafana/grafana/issues/new?assignees=&labels=type%3A+bug&template=1-bug_report.md